### PR TITLE
2.0 strings and vars

### PR DESCRIPTION
**Pull request by pflarr**

### CODE REVIEW

File: lib/pavilion/pav_config.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                    sub_elem=yc.StrElem(),
+                    help_text="Paths to search for Pavilion config files. Pavilion configs (other"
+                              "than this core config) are searched in the given order. In the case"
+                              "of identically named files, directories listed earlier take "
+                              "precedent.")
+    ]
+
+
+PAV_CONFIG_SEARCH_DIRS = [
+    os.environ.get(['PAV_CONFIG_DIR'], './'),
```
Comment (nicholas-sly) [inline lib/pavilion/pav_config.py]: Shouldn't the user's configs override the installation-specific configs?

File: lib/pavilion/schedulers/old_testconfig.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
+class TestConfig:
+
+    def __init__(self, config, var_man=None):
+        """
+        :param config:
```
Comment (nicholas-sly) [inline lib/pavilion/schedulers/old_testconfig.py]: Is this documentation incomplete?

File: lib/pavilion/schedulers/old_testconfig.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    def scheduler_section(self):
+
+        self._config.get('scheduler')
+
+
+
+    def get_result_locations(self):
+        rl = []
+        for k, v in self.ecf.iteritems():
+            if not isinstance(v, dict):
```
Comment (nicholas-sly) [inline lib/pavilion/schedulers/old_testconfig.py]: This seems to not do anything...

File: lib/pavilion/config.py [hunk 4]
```diff
@@ -1,157 +1,309 @@
...
+# Config file types
+CONF_HOST = 'hosts'
+CONF_MODE = 'modes'
+CONF_TEST = 'tests'
+
+def find_config(pav_config, conf_type, conf_name):
+    """Search all of the known configuration directories for a config of the given
+    type and name.
+    :param pav_config:
+    :param conf_type:
```
Comment (nicholas-sly) [inline lib/pavilion/config.py]: Documentation.

File: lib/pavilion/config.py [hunk 5]
```diff
@@ -1,157 +1,309 @@
...
+            return path
+
+    return None
+
+
+def get_tests(pav_config, host, modes, tests):
+    """Get a dictionary of the test configs given a host, list of modes,
+    and a list of tests.
+    :param pav_config:
+    :param Union(str, None) host:
```
Comment (nicholas-sly) [inline lib/pavilion/config.py]: Documentation.

File: lib/pavilion/config.py [hunk 6]
```diff
@@ -1,157 +1,309 @@
...
+class TestConfigurator(object):
+    """
+    """
+
+    def __init__(self, test_cfg, pav_vars, sys_vars, sched_vars):
+        """Create a new TestConfig object.
+        :param dict test_cfg: The raw test configuration.
+        :param dict pav_vars: The pavilion provided variables.
+        :param dict sys_vars: The system provided variables.
+        :param dict sched_vars: A
```
Comment (nicholas-sly) [inline lib/pavilion/config.py]: Documentation?

File: lib/pavilion/pav_config.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    PAV_CONFIG_SEARCH_DIRS.append(os.environ['PAV_CONFIG_DIR'])
+PAV_CONFIG_SEARCH_DIRS.extend([
+    '/etc/pavilion',
+    '/opt/pavilion',
+])
+
+
+def find_pavilion_config():
+    """Search for a pavilion.yaml configuration file. The first one found is used.
+    Directories are searched in this order: $PAV_CONFIG_DIR, $USER/.pavilion, /etc/pavilion/,
```
Comment (nicholas-sly) [inline lib/pavilion/pav_config.py]: Update to new order.

File: lib/pavilion/string_parser.py [hunk 8]
```diff
@@ -0,0 +1,460 @@
...
+# describing the format:
+
+# # pav strings can contain variable references, escapes, or sub strings.
+# PAV_STR  -> TEXT | TEXT VAR PAV_STR | TEXT ESC PAV_STR | TEXT SUB_STR PAV_STR
+# # Text is anything that doesn't start an escape, sub string, or variable reference
+# # It may also be an empty string.
+# TEXT     -> [^\{[]*
+# # Variable references are a bracket surrounded variable name. These can consist
+# # of a var_set name, variable name, variable index, and sub-variable name. All but the
+# # variable name are optional.
```
Comment (nicholas-sly) [inline lib/pavilion/string_parser.py]: Should the optional variables have specified defaults?

File: lib/pavilion/string_parser.py [hunk 9]
```diff
@@ -0,0 +1,460 @@
...
+        end_str = match.group()
+
+        # Process the non-text token (in most cases) that follows.
+        if end_str == '[':
+            tokens.append(SubStringStartToken(pos, pos+1))
+            pos += 1
+        elif end_str.endswith(']'):
+
+            if len(end_str) == 3:
+                separator = end_str[1]
```
Comment (nicholas-sly) [inline lib/pavilion/string_parser.py]: I may be severely lost, but is the precluding the possibility of a multi-character separator?

File: lib/pavilion/string_parser.py [hunk 10]
```diff
@@ -0,0 +1,460 @@
...
+            raise ScanError("Unknown scanning error at character {}.".format(end_str), pos, pos + 1)
+
+    return tokens
+
+
+class Token(object):
+    def __init__(self, start, end):
+        """Scan the string starting at pos to find the end of this token. Save the matching
+        part, start and end. These type of token may be empty.
+        :param int start:
```
Comment (nicholas-sly) [inline lib/pavilion/string_parser.py]: Documentation.

File: lib/pavilion/string_parser.py [hunk 11]
```diff
@@ -0,0 +1,460 @@
...
+
+
+class PavString(Token):
+    """Provides a tokenized representation of a pavilion string, including summary information
+    about variables used, etc. It is itself the root token of the parse tree."""
+
+    def __init__(self, tokens, is_substr=False):
+        """
+        Tokenize the given pav_string.
+        :param list tokens:
```
Comment (nicholas-sly) [inline lib/pavilion/string_parser.py]: Documentation.

File: lib/pavilion/string_parser.py [hunk 12]
```diff
@@ -0,0 +1,460 @@
...
+                var_set.union(token.variables)
+
+            token = token.next_token
+
+        return var_set
+
+    def get_substr_vars(self, var_man, iter_vars):
+        """
+        :param variables.VariableSetManager var_man:
+        :param dict iter_vars:
```
Comment (nicholas-sly) [inline lib/pavilion/string_parser.py:289]: Documentation.

File: lib/pavilion/string_parser.py [hunk 13]
```diff
@@ -0,0 +1,460 @@
...
+
+            token = token.next_token
+
+        return sorted(local_iter_vars)
+
+    def resolve(self, var_man, iter_vars=None):
+        """
+        :param variables.VariableSetManager var_man:
+        :param dict iter_vars:
+        :return:
```
Comment (nicholas-sly) [inline lib/pavilion/string_parser.py]: Documentation.

File: lib/pavilion/string_parser.py [hunk 14]
```diff
@@ -0,0 +1,460 @@
...
+
+                # Now loop over all our local iter vars in sorted order (they needed to be in
+                # some consistent order).
+                while not done:
+                    part = token.resolve(var_man, iter_vars)
+                    parts.append(part)
+
+                    # If there aren't any local iter vars, we're done now.
+                    if not local_iter_vars:
+                        break
```
Comment (nicholas-sly) [inline lib/pavilion/string_parser.py:360]: done = True?

File: lib/pavilion/string_parser.py [hunk 15]
```diff
@@ -0,0 +1,460 @@
...
+    def __init__(self, var, start, end):
+        super(VariableToken, self).__init__(start, end)
+
+        self.var = var
+
+    def resolve(self, var_man, index=None):
+        """Resolve any variables in this token using the variable manager.
+        :param variables.VariableSetManager var_man:
+        :param int index: The index to force for this variable, when it's being iterated over.
+        :return:
```
Comment (nicholas-sly) [inline lib/pavilion/string_parser.py:430]: Documentation.

File: lib/pavilion/variables.py [hunk 16]
```diff
@@ -0,0 +1,477 @@
...
+        key = '.'.join(key)
+
+        return "Error processing variable key '{}': {}".format(key, self.base_message)
+
+
+class VariableSetManager:
+    """This class manages the various sets of variables, provides complex key based lookups,
+    manages conflict resolution, and so on."""
+
+    # The variable sets, in order of resolution priority.
```
Comment (nicholas-sly) [inline lib/pavilion/variables.py]: Increasing priority?

File: lib/pavilion/variables.py [hunk 17]
```diff
@@ -0,0 +1,477 @@
...
+            err.var_set = name
+            raise err
+
+        self.variable_sets[name] = var_set
+
+    def get_permutations(self, used_per_vars):
+        """For every combination of permutation variables (that were used), return a new
+        varset manager.
+        :param set used_per_vars: A list of permutation variable names that were used.
+        :return:
```
Comment (nicholas-sly) [inline lib/pavilion/variables.py:142]: Documentation.

File: lib/pavilion/string_parser.py (hunk 11 above)
Comment (nicholas-sly) [inline lib/pavilion/string_parser.py:223]: Documentation.

File: lib/pavilion/string_parser.py (hunk 10 above)
Comment (nicholas-sly) [inline lib/pavilion/string_parser.py:196]: Documentation.

File: lib/pavilion/string_parser.py (hunk 9 above)
Comment (nicholas-sly) [inline lib/pavilion/string_parser.py:142]: I may be severely lost, but is the precluding the possibility of a multi-character separator?

File: lib/pavilion/string_parser.py (hunk 8 above)
Comment (nicholas-sly) [inline lib/pavilion/string_parser.py:67]: Should the optional variables have specified defaults?

File: lib/pavilion/schedulers/old_testconfig.py (hunk 3 above)
Comment (nicholas-sly) [inline lib/pavilion/schedulers/old_testconfig.py:21]: This seems to not do anything...

File: lib/pavilion/schedulers/old_testconfig.py (hunk 2 above)
Comment (nicholas-sly) [inline lib/pavilion/schedulers/old_testconfig.py:5]: Documentation.

File: lib/pavilion/pav_config.py (hunk 7 above)
Comment (nicholas-sly) [inline lib/pavilion/pav_config.py]: Update to new order.

File: lib/pavilion/config.py (hunk 6 above)
Comment (nicholas-sly) [inline lib/pavilion/config.py]: Documentation?

File: lib/pavilion/config.py (hunk 5 above)
Comment (nicholas-sly) [inline lib/pavilion/config.py]: Documentation.

File: lib/pavilion/config.py (hunk 4 above)
Comment (nicholas-sly) [inline lib/pavilion/config.py]: Documentation.

**Review by nicholas-sly — CHANGES_REQUESTED** [review]
Minor nitpicks and questions.

File: lib/pavilion/string_parser.py (hunk 14 above)
Comment (pflarr) [reply lib/pavilion/string_parser.py:360]: The alternative is:
```suggestion
                        done = True
                        continue
```

File: lib/pavilion/schedulers/old_testconfig.py (hunk 2 above)
Comment (pflarr) [reply lib/pavilion/schedulers/old_testconfig.py:5]: This old code is around for reference only for now.

File: lib/pavilion/config.py (hunk 6 above)
Comment (pflarr) [reply lib/pavilion/config.py]: I shouldn't have checked this file in yet. I wasn't anywhere near done.

**Pull request merged**